flutter: 3.3.8 -> 3.7.9

algorithmiker opened the pull request:
This updates flutter and will also probably fix the dependency- armageddon we are seeing with #211606.

This is mostly @FlafyDev's work, who said it's okay for me to upstream the changes.

Flutter 3.7 was attempted earlier by @AlexandraSjel, whom we thank greatly.

This PR therefore
Closes #215754
Closes #215537
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

AtaraxiaSjel commented:
@gilice, thank you for the work! Unfortunately, I haven't had a lot of free time lately and I'm glad someone took on this problem :) Should I close my PR now?

FlafyDev commented:
> @FlafyDev can you here state publicly that you are ok with those changes being merged? Did you use any work from AtaraxiaSjel?

Yes, I have no problem for these changes to be merged to Nixpkgs.
Thanks for upstreaming them! 

Also no, I haven't used any of the changes from @AtaraxiaSjel's PR.

NickCao commented:
The patches are failing to apply
```
       > applying patch /nix/store/9671j6kqrbyrpzpas7rr172kk7jwbkj5-remove-test-dependency.patch
       > patching file packages/flutter_tools/pubspec.yaml
       > Hunk #1 FAILED at 98.
       > 1 out of 1 hunk FAILED -- saving rejects to file packages/flutter_tools/pubspec.yaml.re
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2022

FlafyDev commented on maintainers/maintainer-list.nix:
```diff
@@ -5022,6 +5022,11 @@
     github = "fkautz";
     githubId = 135706;
   };
+  FlafyDev = {
```
Can you add `name = "Flafy Arazi";`?

Seems as though the CI fails because there is no `name`.

NickCao commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -56,21 +59,36 @@ let
+      rm -r bin/cache/dart-sdk
+    '';
+
+    postFixup = ''
```
Just use `wrapProgram`?

FlafyDev replied:
Originally I did try using `wrapProgram`, but didn't succeed as it behaved unexpectedly(such as pointing to incorrect files) and didn't wrap the program.
Additionally, this line `export PUB_CACHE=\''${PUB_CACHE:-"\$HOME/.pub-cache"}\` might be difficult to add with `wrapProgram`.

Maybe someone else will have better luck with `wrapProgram`, but that's my experience..

MGlolenstine commented:
I receive an error while trying to build for Android.
<details>

<summary>Error</summary>

```
✦ ❯ flutter run
The new artifact "libimobiledevice" was downloaded, but Flutter failed to update its stamp file, receiving the error "FileSystemException: Cannot open file, path =
'/nix/store/6kg35py06f1jvnmglm51nxchy1gcbc3y-flutter-3.7.9-unwrapped/bin/cache/libimobiledevice.stamp' (OS Error: Read-only file system, errno = 30)". Flutter can continue, but the artifact may be re-downloaded on subsequent invocations
until the problem is resolved.
The new artifact "usbmuxd" was downloaded, but Flutter failed to update its stamp file, receiving the error "FileSystemException: Cannot open file, path =
'/nix/store/6kg35py06f1jvnmglm51nxchy1gcbc3y-flutter-3.7.9-unwrapped/bin/cache/usbmuxd.stamp' (OS Error: Read-only file system, errno = 30)". Flutter can continue, but the artifact may be re-downloaded on subsequent invocations until the
problem is resolved.
Running "flutter pub get" in flutter_tools...
Resolving dependencies in ../../../../nix/store/6kg35py06f1jvnmglm51nxchy1gcbc3y-flutter-3.7.9-unwrapped/packages/flutter_tools... (1.6s)
  _fe_analyzer_shared 50.0.0 (58.0.0 available)
  analyzer 5.2.0 (5.10.0 available)
  archive 3.3.2 (3.3.6 available)
  args 2.3.1 (2.4.0 available)
  async 2.10.0 (2.11.0 available)
  built_value 8.4.2 (8.4.4 available)
  checked_yaml 2.0.1 (2.0.2 available)
  collection 1.17.0 (1.17.1 available)
  completion 1.0.0 (1.0.1 available)
  coverage 1.6.1 (1.6.3 available)
  dds 2.5.0 (2.7.7 available)
  dds_service_extensions 1.3.1 (1.3.3 available)
  devtools_shared 2.18.0 (2.22.2 available)
  dwds 16.0.2+1 (18.0.2 available)
  fixnum 1.0.1 (1.1.0 available)
  frontend_server_client 3.1.0 (3.2.0 available)
  html 0.15.1 (0.15.2 available)
  intl 0.17.0 (0.18.0 available)
  io 1.0.3 (1.0.4 available)
  js 0.6.5 (0.6.7 available)
  json_annotation 4.7.0 (4.8.0 available)
  logging 1.1.0 (1.1.1 available)
  matcher 0.12.13 (0.12.15 available)
  meta 1.8.0 (1.9.1 available)
  mime 1.0.2 (1.0.4 available)
  multicast_dns 0.3.2+2 (0.3.2+3 available)
  native_stack_traces 0.5.2 (0.5.5 available)
  node_preamble 2.0.1 (2.0.2 available)
  path 1.8.2 (1.8.3 available)
  petitparser 5.1.0 (5.3.0 available)
  pubspec_parse 1.2.1 (1.2.2 available)
  source_maps 0.10.11 (0.10.12 available)
  sse 4.1.1 (4.1.2 available)
  test 1.22.0 (1.24.1 available)
  test_api 0.4.16 (0.5.1 available)
  test_core 0.4.20 (0.5.1 available)
  vm_service 9.4.0 (11.3.0 available)
  web_socket_channel 2.2.0 (2.3.0 available)
  webdriver 3.0.1 (3.0.2 available)
Got dependencies in ../../../../nix/store/6kg35py06f1jvnmglm51nxchy1gcbc3y-flutter-3.7.9-unwrapped/packages/flutter_tools!
Cannot open file, path = '../../../../nix/store/6kg35py06f1jvnmglm51nxchy1gcbc3y-flutter-3.7.9-unwrapped/packages/flutter_tools/pubspec.lock' (OS Error: Read-only file system, errno = 30)
pub get failed
command: "/nix/store/6kg35py06f1jvnmglm51nxchy1gcbc3y-flutter-3.7.9-unwrapped/bin/cache/dart-sdk/bin/dart __deprecated_pub --color --directory
../../../../nix/store/6kg35py06f1jvnmglm51nxchy1gcbc3y-flutter-3.7.9-unwrapped/packages/flutter_tools get --example"
pub env: {
  "FLUTTER_ROOT": "/nix/store/6kg35py06f1jvnmglm51nxchy1gcbc3y-flutter-3.7.9-unwrapped",
  "PUB_ENVIRONMENT": "flutter_cli:get",
  "PUB_CACHE": "/home/arch/.pub-cache",
}
exit code: 66
```
</details>

I have added the above repository to the flake, locked it to the latest commit
```nix
    flutterNixpkgs = {
      type = "github";
      owner = "gilice";
      repo = "nixpkgs";
      rev = "dd242357fde311d4695d265baebd232a7ad7cf1a";
    };
```
And I've added `flutterPkgs.flutter` to the `buildInputs` (for the `mkShell`).

FlafyDev commented:
> also @FlafyDev  flutter2 fails because it's dart doesn't have precache, we need separate builders.

Why package flutter 2?

FlafyDev commented:
> @FlafyDev `firmware-updater` depends on it.

Latest version of `firmware-updater` is flutter 3.
Same thing with `fluffychat`

MGlolenstine commented:
The `flutter run` still fails for me.
It still tries to write `/bin/cache/artifacts` into the nix store.
<details>
<summary>`flutter run` error log</summary>

```
✦ ❯ flutter run

Oops; flutter has exited unexpectedly: "FileSystemException: Creation failed, path = '/nix/store/jnwvgnqxabsfpja8440i95iybx4y260b-flutter-unwrapped-3.7.9/bin/cache/artifacts' (OS Error: Read-only file system, errno = 30)".
A crash report has been written to /home/arch/dartProjects/project/flutter_02.log.
This crash may already be reported. Check GitHub for similar crashes.
https://github.com/flutter/flutter/issues?q=is%3Aissue+FileSystemException%3A+Creation+failed%2C+path+%3D+%27%2Fnix%2Fstore%2Fjnwvgnqxabsfpja8440i95iybx4y260b-flutter-unwrapped-3.7.9%2Fbin%2Fcache%2Fartifacts%27+%28OS+Error%3A+Read-only+file+system%2C+errno+%3D+30%29

To report your crash to the Flutter team, first read the guide to filing a bug.
https://flutter.dev/docs/resources/bug-reports

Create a new GitHub issue by pasting this link into your browser and completing the issue template. Thank you!
https://github.com/flutter/flutter/issues/new?title=%5Btool_crash%5D+FileSystemException%3A+Creation+failed%2C+OS+Error%3A+Read-only+file+system%2C+errno+%3D+30&body=%23%23+Command%0A%60%60%60%0Aflutter+--no-version-check+run%0A%60%60%60%0A%0A%23%23+Steps+to+Reproduce%0A1.+...%0A2.+...%0A3.+...%0A%0A%23%23+Logs%0AFileSystemException%3A+Creation+failed%2C+OS+Error%3A+Read-only+file+system%2C+errno+%3D+30%0A%60%60%60%0A%230++++++_Directory.createSync+%28dart%3Aio%2Fdirectory_impl.dart%3A129%3A7%29%0A%231++++++ForwardingDirectory.createSync+%28package%3Afile%2Fsrc%2Fforwarding%2Fforwarding_directory.dart%3A20%3A16%29%0A%232++++++ErrorHandlingDirectory.createSync.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Ferror_handling_io.dart%3A434%3A22%29%0A%233++++++_runSync+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Ferror_handling_io.dart%3A600%3A14%29%0A%234++++++ErrorHandlingDirectory.createSync+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Ferror_handling_io.dart%3A433%3A12%29%0A%235++++++Cache.getCacheDir+%28package%3Aflutter_tools%2Fsrc%2Fcache.dart%3A555%3A11%29%0A%236++++++Cache.getCacheArtifacts+%28package%3Aflutter_tools%2Fsrc%2Fcache.dart%3A565%3A36%29%0A%237++++++Cache.getArtifactDirectory+%28package%3Aflutter_tools%2Fsrc%2Fcache.dart%3A573%3A12%29%0A%238++++++new+FuchsiaArtifacts.find+%28package%3Aflutter_tools%2Fsrc%2Ffuchsia%2Ffuchsia_sdk.dart%3A122%3A42%29%0A%239++++++runInContext.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fcontext_runner.dart%3A230%3A48%29%0A%2310+++++AppContext._generateIfNecessary.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A104%3A42%29%0A%2311+++++_LinkedHashMapMixin.putIfAbsent+%28dart%3Acollection-patch%2Fcompact_hash.dart%3A535%3A23%29%0A%2312+++++AppContext._generateIfNecessary+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A92%3A20%29%0A%2313+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A121%3A32%29%0A%2314+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A119%3A24%29%0A%2315+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A119%3A24%29%0A%2316+++++fuchsiaArtifacts+%28package%3Aflutter_tools%2Fsrc%2Fglobals.dart%3A65%3A51%29%0A%2317+++++runInContext.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fcontext_runner.dart%3A236%3A35%29%0A%2318+++++AppContext._generateIfNecessary.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A104%3A42%29%0A%2319+++++_LinkedHashMapMixin.putIfAbsent+%28dart%3Acollection-patch%2Fcompact_hash.dart%3A535%3A23%29%0A%2320+++++AppContext._generateIfNecessary+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A92%3A20%29%0A%2321+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A121%3A32%29%0A%2322+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A119%3A24%29%0A%2323+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A119%3A24%29%0A%2324+++++fuchsiaWorkflow+%28package%3Aflutter_tools%2Fsrc%2Ffuchsia%2Ffuchsia_workflow.dart%3A12%3A49%29%0A%60%60%60%0A%60%60%60%0Aencountered+exception%3A+FileSystemException%3A+Creation+failed%2C+path+%3D+%27%2Fnix%2Fstore%2Fjnwvgnqxabsfpja8440i95iybx4y260b-flutter-unwrapped-3.7.9%2Fbin%2Fcache%2Fartifacts%27+%28OS+Error%3A+Read-only+file+system%2C+errno+%3D+30%29%0A%0A%230++++++_Directory.createSync+%28dart%3Aio%2Fdirectory_impl.dart%3A129%3A7%29%0A%231++++++ForwardingDirectory.createSync+%28package%3Afile%2Fsrc%2Fforwarding%2Fforwarding_directory.dart%3A20%3A16%29%0A%232++++++ErrorHandlingDirectory.createSync.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Ferror_handling_io.dart%3A434%3A22%29%0A%233++++++_runSync+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Ferror_handling_io.dart%3A600%3A14%29%0A%234++++++ErrorHandlingDirectory.createSync+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Ferror_handling_io.dart%3A433%3A12%29%0A%235++++++Cache.getCacheDir+%28package%3Aflutter_tools%2Fsrc%2Fcache.dart%3A555%3A11%29%0A%236++++++Cache.getCacheArtifacts+%28package%3Aflutter_tools%2Fsrc%2Fcache.dart%3A565%3A36%29%0A%237++++++Cache.getArtifactDirectory+%28package%3Aflutter_tools%2Fsrc%2Fcache.dart%3A573%3A12%29%0A%238++++++new+FuchsiaArtifacts.find+%28package%3Aflutter_tools%2Fsrc%2Ffuchsia%2Ffuchsia_sdk.dart%3A122%3A42%29%0A%239++++++runInContext.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fcontext_runner.dart%3A230%3A48%29%0A%2310+++++AppContext._generateIfNecessary.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A104%3A42%29%0A%2311+++++_LinkedHashMapMixin.putIfAbsent+%28dart%3Acollection-patch%2Fcompact_hash.dart%3A535%3A23%29%0A%2312+++++AppContext._generateIfNecessary+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A92%3A20%29%0A%2313+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A121%3A32%29%0A%2314+++++fuchsiaArtifacts+%28package%3Aflutter_tools%2Fsrc%2Fglobals.dart%3A65%3A51%29%0A%2315+++++runInContext.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fcontext_runner.dart%3A236%3A35%29%0A%2316+++++AppContext._generateIfNecessary.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A104%3A42%29%0A%2317+++++_LinkedHashMapMixin.putIfAbsent+%28dart%3Acollection-patch%2Fcompact_hash.dart%3A535%3A23%29%0A%2318+++++AppContext._generateIfNecessary+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A92%3A20%29%0A%2319+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A121%3A32%29%0A%2320+++++fuchsiaWorkflow+%28package%3Aflutter_tools%2Fsrc%2Ffuchsia%2Ffuchsia_workflow.dart%3A12%3A49%29%0A%2321+++++runInContext.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fcontext_runner.dart%3A197%3A26%29%0A%2322+++++AppContext._generateIfNecessary.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A104%3A42%29%0A%2323+++++_LinkedHashMapMixin.putIfAbsent+%28dart%3Acollection-patch%2Fcompact_hash.dart%3A535%3A23%29%0A%2324+++++AppContext._generateIfNecessary+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A92%3A20%29%0A%2325+++++AppContext.get+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A121%3A32%29%0A%2326+++++deviceManager+%28package%3Aflutter_tools%2Fsrc%2Fglobals.dart%3A55%3A45%29%0A%2327+++++_DefaultDoctorValidatorsProvider.validators+%28package%3Aflutter_tools%2Fsrc%2Fdoctor.dart%3A166%3A19%29%0A%2328+++++Doctor.validators+%28package%3Aflutter_tools%2Fsrc%2Fdoctor.dart%3A229%3A47%29%0A%2329+++++Doctor.startValidatorTasks+%28package%3Aflutter_tools%2Fsrc%2Fdoctor.dart%3A235%3A45%29%0A%2330+++++DoctorText._validatorTasks+%28package%3Aflutter_tools%2Fsrc%2Fdoctor.dart%3A739%3A60%29%0A%2331+++++DoctorText._validatorTasks+%28package%3Aflutter_tools%2Fsrc%2Fdoctor.dart%29%0A%2332+++++DoctorText._runDiagnosis+%28package%3Aflutter_tools%2Fsrc%2Fdoctor.dart%3A743%3A53%29%0A%2333+++++DoctorText.piiStrippedText+%28package%3Aflutter_tools%2Fsrc%2Fdoctor.dart%3A736%3A47%29%0A%2334+++++DoctorText.piiStrippedText+%28package%3Aflutter_tools%2Fsrc%2Fdoctor.dart%29%0A%2335+++++CrashReporter.informUser+%28package%3Aflutter_tools%2Fsrc%2Freporting%2Fcrash_reporting.dart%3A92%3A32%29%0A%2336+++++_handleToolError+%28package%3Aflutter_tools%2Frunner.dart%3A169%3A36%29%0A%3Casynchronous+suspension%3E%0A%2337+++++AppContext.run.%3Canonymous+closure%3E+%28package%3Aflutter_tools%2Fsrc%2Fbase%2Fcontext.dart%3A150%3A19%29%0A%3Casynchronous+suspension%3E%0A%2338+++++main+%28package%3Aflutter_tools%2Fexecutable.dart%3A91%3A3%29%0A%3Casynchronous+suspension%3E%0A%0A%60%60%60%0A%0A%23%23+Flutter+Application+Metadata%0A%2A%2AType%2A%2A%3A+malformed%0A%2A%2AVersion%2A%2A%3A+1.0.2%2B3%0A%2A%2AMaterial%2A%2A%3A+true%0A%2A%2AAndroid+X%2A%2A%3A+false%0A%2A%2AModule%2A%2A%3A+false%0A%2A%2APlugin%2A%2A%3A+false%0A%2A%2AAndroid+package%2A%2A%3A+null%0A%2A%2AiOS+bundle+identifier%2A%2A%3A+null%0A%2A%2ACreation+channel%2A%2A%3A+null%0A%2A%2ACreation+framework+version%2A%2A%3A+null%0A%23%23%23+Plugins%0Afile_picker-5.2.6%0Aflutter_archive-5.0.0%0Aflutter_plugin_android_lifecycle-2.0.9%0Afluttertoast-8.2.1%0Amcumgr_flutter-0.1.1%2B1%0Anordic_dfu-6.0.1%0Apath_provider-2.0.13%0Apath_provider_android-2.0.24%0Apath_provider_foundation-2.1.3%0Apath_provider_linux-2.1.10%0Apath_provider_windows-2.1.5%0Apermission_handler-10.2.0%0Apermission_handler_android-10.2.0%0Apermission_handler_apple-9.0.7%0Apermission_handler_windows-0.1.2%0Aplugin%0Aprinting-5.10.1%0Aquick_blue%0Ashare_plus-6.3.1%0Ashared_preferences-2.0.18%0Ashared_preferences_android-2.0.17%0Ashared_preferences_foundation-2.1.5%0Ashared_preferences_linux-2.1.5%0Ashared_preferences_web-2.0.6%0Ashared_preferences_windows-2.1.5%0Asqflite-2.2.5%0Aurl_launcher_linux-3.0.4%0Aurl_launcher_web-2.0.16%0Aurl_launcher_windows-3.0.5%0A%0A&labels=tool%2Csevere%3A+crash
```
</details>

MGlolenstine commented:
I still got some issues on my side.
<details>
<summary>`flutter run`</summary>

```
✦ ❯ flutter run
Launching lib/main.dart on Pixel 7 Pro in debug mode...

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processDebugResources'.
> A failure occurred while executing com.android.build.gradle.internal.res.LinkApplicationAndroidResourcesTask$TaskAction
   > AAPT2 aapt2-7.4.1-8841542-linux Daemon #0: Daemon startup failed
     This should not happen under normal circumstances, please file an issue if it does.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 20s
Running Gradle task 'assembleDebug'...                             20.7s
Exception: Gradle task assembleDebug failed with exit code 1
```

</details>

<details>
<summary>`flutter build apk`</summary>

```
✦ ❯ flutter build apk
Running "flutter pub get" in cst_mobile_manufacturer...
Resolving dependencies... (1.7s)
  _fe_analyzer_shared 52.0.0 (58.0.0 available)
  analyzer 5.4.0 (5.10.0 available)
  async 2.10.0 (2.11.0 available)
  characters 1.2.1 (1.3.0 available)
  charts_flutter 0.12.0 (discontinued)
  cli_util 0.3.5 (0.4.0 available)
  collection 1.17.0 (1.17.1 available)
  dart_code_metrics 5.6.0 (5.7.0 available)
  dart_code_metrics_presets 1.4.1 (1.5.0 available)
  dart_style 2.2.5 (2.3.0 available)
  ffigen 7.2.7 (7.2.10 available)
  intl 0.17.0 (0.18.0 available)
  js 0.6.5 (0.6.7 available)
  matcher 0.12.13 (0.12.15 available)
  material_color_utilities 0.2.0 (0.3.0 available)
  meta 1.8.0 (1.9.1 available)
  path 1.8.2 (1.8.3 available)
  path_provider 2.0.13 (2.0.14 available)
  path_provider_foundation 2.1.3 (2.2.1 available)
  pdf 3.9.0 (3.10.1 available)
  permission_handler_apple 9.0.7 (9.0.8 available)
  petitparser 5.1.0 (5.3.0 available)
  pointycastle 3.6.2 (3.7.2 available)
  printing 5.10.1 (5.10.3 available)
  rxdart 0.26.0 (0.27.7 available)
  shared_preferences 2.0.18 (2.1.0 available)
  shared_preferences_android 2.0.17 (2.1.0 available)
  shared_preferences_foundation 2.1.5 (2.2.0 available)
  shared_preferences_linux 2.1.5 (2.2.0 available)
  shared_preferences_platform_interface 2.1.1 (2.2.0 available)
  shared_preferences_web 2.0.6 (2.1.0 available)
  shared_preferences_windows 2.1.5 (2.2.0 available)
  sqflite 2.2.5 (2.2.6 available)
  sqflite_common 2.4.2+2 (2.4.3 available)
  test_api 0.4.16 (0.5.1 available)
  win32 3.1.3 (4.1.1 available)
Got dependencies!

💪 Building with sound null safety 💪


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeReleaseResources'.
> Multiple task action failures occurred:
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.4.1-8841542-linux Daemon #0: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.4.1-8841542-linux Daemon #1: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.4.1-8841542-linux Daemon #2: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.4.1-8841542-linux Daemon #3: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.4.1-8841542-linux Daemon #4: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.4.1-8841542-linux Daemon #6: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.4.1-8841542-linux Daemon #5: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.4.1-8841542-linux Daemon #7: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 44s
Running Gradle task 'assembleRelease'...                           45.4s
Gradle task assembleRelease failed with exit code 1
```

</details>

I don't know if this error is related to Flutter, but I haven't gotten it with the default installation (Arch Linux through AUR, now on NixOS).

MGlolenstine commented:
Is there an option to add Android support, as that's what I'm using flutter for.

MGlolenstine commented:
It does seem to build for `linux`.
(ignore warnings and runtime errors, as the project is strictly android only.)

<details>

```
✦ ❯ flutter build linux
Running "flutter pub get" in project...
Resolving dependencies... (1.9s)
  _fe_analyzer_shared 52.0.0 (58.0.0 available)
  analyzer 5.4.0 (5.10.0 available)
  async 2.10.0 (2.11.0 available)
  characters 1.2.1 (1.3.0 available)
  charts_flutter 0.12.0 (discontinued)
  cli_util 0.3.5 (0.4.0 available)
  collection 1.17.0 (1.17.1 available)
  dart_code_metrics 5.6.0 (5.7.0 available)
  dart_code_metrics_presets 1.4.1 (1.5.0 available)
  dart_style 2.2.5 (2.3.0 available)
  ffigen 7.2.7 (7.2.10 available)
  intl 0.17.0 (0.18.0 available)
  js 0.6.5 (0.6.7 available)
  matcher 0.12.13 (0.12.15 available)
  material_color_utilities 0.2.0 (0.3.0 available)
  meta 1.8.0 (1.9.1 available)
  path 1.8.2 (1.8.3 available)
  path_provider 2.0.13 (2.0.14 available)
  path_provider_foundation 2.1.3 (2.2.1 available)
  pdf 3.9.0 (3.10.1 available)
  permission_handler_apple 9.0.7 (9.0.8 available)
  petitparser 5.1.0 (5.3.0 available)
  pointycastle 3.6.2 (3.7.2 available)
  printing 5.10.1 (5.10.3 available)
  rxdart 0.26.0 (0.27.7 available)
  shared_preferences 2.0.18 (2.1.0 available)
  shared_preferences_android 2.0.17 (2.1.0 available)
  shared_preferences_foundation 2.1.5 (2.2.0 available)
  shared_preferences_linux 2.1.5 (2.2.0 available)
  shared_preferences_platform_interface 2.1.1 (2.2.0 available)
  shared_preferences_web 2.0.6 (2.1.0 available)
  shared_preferences_windows 2.1.5 (2.2.0 available)
  sqflite 2.2.5 (2.2.6 available)
  sqflite_common 2.4.2+2 (2.4.3 available)
  test_api 0.4.16 (0.5.1 available)
  win32 3.1.3 (4.1.1 available)
Got dependencies!

💪 Building with sound null safety 💪

Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
CMake Warning (dev) at /nix/store/3ca5j5y66apri335b468qsd3vqnn381k-cmake-3.25.1/share/cmake-3.25/Modules/ExternalProject.cmake:3075 (message):
  The DOWNLOAD_EXTRACT_TIMESTAMP option was not given and policy CMP0135 is
  not set.  The policy's OLD behavior will be used.  When using a URL
  download, the timestamps of extracted files should preferably be that of
  the time of extraction, otherwise code that depends on the extracted
  contents might not be rebuilt if the URL changes.  The OLD behavior
  preserves the timestamps from the archive instead, but this is usually not
  what you want.  Update your project to the NEW behavior or specify the
  DOWNLOAD_EXTRACT_TIMESTAMP option with a value of true to avoid this
  robustness issue.
Call Stack (most recent call first):
  /nix/store/3ca5j5y66apri335b468qsd3vqnn381k-cmake-3.25.1/share/cmake-3.25/Modules/ExternalProject.cmake:4185 (_ep_add_download_command)
  CMakeLists.txt:9 (ExternalProject_Add)
This warning is for project developers.  Use -Wno-dev to suppress it.

Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Building Linux application...

project on  new_flake_test [$!?] is 📦 v0.1.0 via 🎯 v2.19.6 via 🦀 via ❄️  impure (nix-shell) took 43s
✦ ❯ flutter run -d linux
Launching lib/main.dart on Linux in debug mode...
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
CMake Warning (dev) at /nix/store/3ca5j5y66apri335b468qsd3vqnn381k-cmake-3.25.1/share/cmake-3.25/Modules/ExternalProject.cmake:3075 (message):
  The DOWNLOAD_EXTRACT_TIMESTAMP option was not given and policy CMP0135 is
  not set.  The policy's OLD behavior will be used.  When using a URL
  download, the timestamps of extracted files should preferably be that of
  the time of extraction, otherwise code that depends on the extracted
  contents might not be rebuilt if the URL changes.  The OLD behavior
  preserves the timestamps from the archive instead, but this is usually not
  what you want.  Update your project to the NEW behavior or specify the
  DOWNLOAD_EXTRACT_TIMESTAMP option with a value of true to avoid this
  robustness issue.
Call Stack (most recent call first):
  /nix/store/3ca5j5y66apri335b468qsd3vqnn381k-cmake-3.25.1/share/cmake-3.25/Modules/ExternalProject.cmake:4185 (_ep_add_download_command)
  CMakeLists.txt:9 (ExternalProject_Add)
This warning is for project developers.  Use -Wno-dev to suppress it.

Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Package libdeflate was not found in the pkg-config search path.
Perhaps you should add the directory containing `libdeflate.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libdeflate', required by 'libtiff-4', not found
Building Linux application...
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: UnimplementedError: getExternalStoragePath() has not been implemented.
#0      PathProviderPlatform.getExternalStoragePath (package:path_provider_platform_interface/path_provider_platform_interface.dart:69:5)
#1      getExternalStorageDirectory (package:path_provider/path_provider.dart:137:40)
#2      DatabaseUtils.OpenDatabase (package:project/database_objects/database_utils.dart:121:34)
#3      DatabaseUtils.results (package:project/database_objects/database_utils.dart:89:30)
#4      main (package:project/main.dart:77:46)
<asynchronous suspension>

Syncing files to device Linux...                                   197ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

💪 Running with sound null safety 💪

An Observatory debugger and profiler on Linux is available at: http://127.0.0.1:39031/j-kH-pG3FLA=/
The Flutter DevTools debugger and profiler on Linux is available at: http://127.0.0.1:9100?uri=http://127.0.0.1:39031/j-kH-pG3FLA=/
flutter: Set rid.debugLock to change if/how locking the rid store is logged
flutter: Set rid.debugReply to change if/how posted replies are logged
flutter: Set rid.replyTimeout to change the default for if/when messages without reply time out

Application finished.
```
</details>

MGlolenstine commented:
> Is there an option to add Android support, as that's what I'm using flutter for.
> 
> (I am currently packaging a custom Android SDK. Is there anything else I'd need for a working Android setup with this derivation?)

To answer myself, adding `pkgs.androidenv.composeAndroidPackages` and the following line into the flake, worked.
```
GRADLE_OPTS = "-Dorg.gradle.project.android.aapt2FromMavenOverride=${androidSdk}/libexec/android-sdk/build-tools/${buildToolsVersion}/aapt2";
```

MGlolenstine commented:
@gilice I currently have it working, so maybe for the future, but thanks for the link!
My current flake doesn't have many more lines than the configuration needed for the `dart-flutter-nix`, but it's mostly the lower-end compatibility that I have to patch. (tons of native deps).

algorithmiker committed:
add FlafyDev to maintainers

algorithmiker committed:
flutter: 3.3.8 -> 3.7.9; small refactor

This updates flutter and will also probably fix the dependency-
armageddon we are seeing with #211606.

This is based on @FlafyDev's work, who said it's okay for me to upstream
the changes.

Flutter 3.7 was attempted earlier by @AtaraxiaSjel, whom we thank
greatly.

This PR therefore
Closes #215754
Closes #215537

Co-authored-by: FlafyDev <44374434+flafydev@users.noreply.github.com>
Co-authored-by: AtaraxiaSjel<5314145+ataraxiasjel@users.noreply.github.com>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2041

NickCao commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -55,37 +71,67 @@ let
       export STAMP_PATH="$FLUTTER_ROOT/bin/cache/flutter_tools.stamp"
 
-      export DART_SDK_PATH="${dart}"
+      export DART_SDK_PATH="$(which dart)"
```
Shouldn't it point to the root of the dart sdk, not the dart binary?

NickCao commented on pkgs/development/compilers/flutter/default.nix:
```diff
@@ -19,37 +18,34 @@ let
 
-  v2 = flutterDrv {
-    pname = "flutter";
+  v2 = mkFlutter {
```
The only consumer of flutter2 in nixpkgs, firmware-updater does seem to support flutter3: https://github.com/canonical/firmware-updater/blob/6e7dbdb64e344633ea62874b54ff3990bd3b8440/pubspec.yaml#L7, Shall we just drop flutter2?

NickCao commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -55,37 +71,67 @@ let
+
     doInstallCheck = true;
-    nativeInstallCheckInputs = [ which ];
+    installCheckInputs = [ which git ];
```
These binaries are to be executed on the build platform in install check phase, they belong to `nativeInstallCheckInputs`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/900

NickCao commented:
Result of `nixpkgs-review pr 224090` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>firmware-updater</li>
    <li>yubioath-flutter</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>flutter (flutterPackages.stable)</li>
    <li>flutter2 (flutterPackages.v2)</li>
    <li>hover</li>
  </ul>
</details>

Failures are due to hash mismatch.

NickCao commented:
Result of `nixpkgs-review pr 224090` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>yubioath-flutter</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>firmware-updater</li>
    <li>flutter (flutterPackages.stable)</li>
    <li>flutter2 (flutterPackages.v2)</li>
    <li>hover</li>
  </ul>
</details>

There's still some issue with the hashes.

algorithmiker committed:
yubioath-flutter: fix hashes for flutter 3.7.9

NickCao commented:
Failed once again due to hash mismatch
```
error: hash mismatch in fixed-output derivation '/nix/store/483xb8c5mx76r65rjr98nvi31f64zv14-yubioath-flutter-6.1.0-deps-flutter-v3.7.9-aarch64-linux.tar.gz.drv':
         specified: sha256-gReSoBt+mjtQHLEBdWwJuYClv1wG3NLRIpJZa3Mcixo=
            got:    sha256-5TuZn8GNVKIk5J5MHyuqMScHZ3WqEOVI52Zbfd/1EKk=
```

hacker1024 commented:
Would you be able to split this up into smaller commits? There is a lot being changed here (preload support, wrapper changes, added libraries, etc.).

hacker1024 commented:
Regarding the changes to wrapping, the FHS, and libraries, please consider my alternative approach over at #201027.

This PR makes use of `autoPatchelfHook` to patch SDK artifacts. My solution instead [wraps the `flutter` tool](https://github.com/NixOS/nixpkgs/blob/341fa709ffca3d4c850e64bbbaec5d2a93c234fb/pkgs/development/compilers/flutter/wrapper.nix#L93) to set appropriate path environment variables. Here is my reasoning supporting this method:

- To my understanding, there's no guarantee that all the libraries used for Linux desktop support will be present in the SDK at the time of patching. At any point, the `flutter` tool may decide to download or update libraries, breaking compilation and execution.
I think it will be more appropriate to build native binaries within Nixpkgs, rather than patch the binaries from the SDK. This is being done in #212328.
- For proper Linux desktop support, some environment variables such as `PATH` and `PKG_CONFIG_PATH` must be set. It makes sense to specify all the other build dependencies in the wrapper as well.

As for the differing wrapping methods, I have a [function](https://github.com/NixOS/nixpkgs/blob/341fa709ffca3d4c850e64bbbaec5d2a93c234fb/pkgs/development/compilers/flutter/sdk-symlink.nix) that builds a derivation composing a single modified `flutter` executable with an existing SDK directory.

- This allows arbitrary modifications to be made easily to the `flutter` executable, such as those made with `makeWrapper` or `buildFHSUserEnv`, in a composable manner. I use this to provide a raw "wrapped" version of the SDK, and then again after putting that wrapped executable in a FHS environment for compatibility with some of Flutter build tools. This ensures that the `flutter` executable is always inside a complete SDK, regardless of the environment at runtime.
- Modifying the `flutter` script directly, as I understand this PR does, is depending on factors we do not control, and is prone to breakage. What if the SDK later switches to a different scripting language, or a binary?

My PR also makes it very clear what is required only for Linux desktop support, and allows the feature to be disabled to reduce closure size.

I have last tested my PR with Flutter 3.7.3, and it works nicely. I am now attempting to add the new preloading functionality for support with 3.7.9+.

hacker1024 commented:
Sure. I'm looking forward to fixing reproducibility, eager to see what you come up with on that front. I'll implement 3.7.10 on my branch in the next day or two.

FlafyDev commented:
Hey
I'd like to clear up what this PR changes.

If I haven't missed anything, this PR updates Flutter as the title says. But it also updates the way Flutter gets its artifacts.

Usually, Flutter(in nixpkgs and in general) downloads all the artifacts at runtime when you first run `flutter`. It will also update them whenever it deems necessary.

This PR, on the other hand, downloads and packages these artifacts with the derivation and disables Flutter's ability to modify/update them.

I think this approach is preferable to downloading the artifacts at runtime for the following reasons:

- Packaging Flutter apps will be far more convenient since Flutter can't download different artifact versions to the vendor derivation.
- Flutter will attempt to download different artifacts on different versions. So if you regularly use different versions of Flutter from nixpkgs(because you are working on multiple projects or something) you'll have to wait for Flutter to redownload all of the artifacts when switching to a different version. (I am pretty sure this is a nix-only problem as the package redirects all of the artifacts to the same place: ~/.cache.) This PR fixes that.
- With the artifacts packaged and elfpatched, Flutter no longer needs an FHS environment (for building to Linux, building to Android will still require FHS)
- In Flutter's releases and packages of other distros, Flutter already comes with all the artifacts. Currently, the Flutter package in nixpkgs deletes them so Flutter can redownload them to ~/.cache.
- The artifacts should be a part of the Flutter package.


> * To my understanding, there's no guarantee that all the libraries used for Linux desktop support will be present in the SDK at the time of patching. At any point, the `flutter` tool may decide to download or update libraries, breaking compilation and execution.

You are referencing two different things here.

There are libraries Flutter needs to be available on the system at runtime for building projects to different platforms. For example, you'll need cmake, ninja, gtk, etc... for building Flutter projects to Linux.

Then there are artifacts, which are files Flutter always needs at runtime and are usually bundled with it when getting it (Nixpkgs is the exception, as I've explained above). This PR makes it so artifacts are bundled with Flutter at its unwrapped derivation (IMO this is necessary. See my reasons above)

The artifacts are what Flutter attempts to download/update at runtime. And the former(libraries, like gtk that Flutter needs to build apps to Linux) are the ones not guaranteed to be available at runtime.

>  I think it will be more appropriate to build native binaries within Nixpkgs, rather than patch the binaries from the SDK. This is being done in flutter-engine: init #212328.

I don't know how more convenient it will be than just using prebuilts, but having the option to build yourself or using prebuilts from nixpkgs rather than Google will be nice, I guess.
This isn't related to the PR, though. Correct me if I'm wrong


---

I do think the suggestion in #201027 is a good idea. That is, having different wrappers applied depending on what you'll use Flutter for. 

For example, currently in nixpkgs unwrapped Flutter is wrapped with many dependencies for building Flutter apps for Android, which is not ideal.

Additionally, after this PR Flutter will no longer require FHS for building to Linux. But FHS will still be needed when building for Android, so having that managed depending on the needs would be nice.

That being said, and if I'm not wrong, this PR isn't focused on how Flutter is wrapped, so I don't think it should interfere with #201027. Perhaps there should be some changes to this PR to minimize the difference in the wrapping code, though..

hacker1024 commented:
Thanks for that breakdown, it's very clear.

I think the easiest thing to do at this point is to incorporate the artifact and patching changes into my branch, and then update the wrappers accordingly. I've already done a lot of the modularisation required to pull out Android and the FHS components.

As for Android, I've been meaning to look into it, but figured that was out of scope for the time being. I think the current Android solution is inappropriate - to my understanding, most of the added libraries are there for the functioning of the Android SDK, but this is beyond the scope of Flutter. It makes more sense to have another derivation to patch or wrap the SDK, as is done in [android-nixpkgs](https://github.com/tadfisher/android-nixpkgs).

cole-h commented:
@ofborg eval
